### PR TITLE
Flat cursors

### PR DESCRIPTION
**Pull request by davidegrohmann**
Have single "high level" cursors that deals with both transaction state and storage.

### CODE REVIEW

**Comment by davidegrohmann** [comment]
[Flatten TxState handling into cursors to reduce  nesting.](https://trello.com/c/e6QNV7pd/12-flatten-txstate-handling-into-cursors-to-reduce-nesting)

**Comment by davidegrohmann** [comment]
@thobe @tinwelint @MishaDemianenko any progress on this PR?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreSingleNodeCursor.java [hunk 1]
```diff
@@ -76,30 +87,61 @@ public NodeItem get()
...
     {
-        labels = null;
+        state = null;
+        clearCurrentState();
+        fetched = false;
         nodeRecord.clear();
         instanceCache.accept( this );
     }
 
+    private void clearCurrentState()
```
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreSingleNodeCursor.java]: currentNodeState

File: community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java [hunk 2]
```diff
@@ -420,6 +420,77 @@ protected boolean fetchNext()
         };
     }
 
+    public static PrimitiveIntSet singletonSet( final int item )
```
Comment (ragadeeshu) [inline community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java]: Is this used?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxAllPropertyCursor.java [hunk 3]
```diff
@@ -42,24 +42,29 @@ public TxAllPropertyCursor( Consumer<TxAbstractPropertyCursor> instanceCache )
...
-                if ( changedProperty != null )
+                if ( state.isPropertyRemoved( propertyKeyId ) )
                 {
-                    this.property = (DefinedProperty) changedProperty;
-                    return true;
+                    continue;
                 }
 
-                if ( !state.isPropertyRemoved( propertyKeyId ) )
+                if ( property == null )
```
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxAllPropertyCursor.java]: Some comments for these null checks, just for clarity

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java [hunk 4]
```diff
@@ -118,6 +124,10 @@ public Object value()
     @Override
     public PropertyItem get()
     {
+        if ( !fetched )
+        {
+            throw new IllegalStateException();
```
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java]: Add a message to the exception

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java [hunk 5]
```diff
@@ -33,27 +37,48 @@
...
+        storagePropertyIterator = state == null ? null : state.addedProperties();
+        initialize( ALWAYS_TRUE_INT, firstPropertyId, lock, state );
         return this;
     }
 
     @Override
-    protected boolean loop()
+    protected boolean loadNextFromDisk()
+    {
+        return fromDisk;
```
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java]: Is there a clearer way to do this? :)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreSinglePropertyCursor.java [hunk 6]
```diff
@@ -21,29 +21,40 @@
...
+        initialize( key -> key == propertyKeyId, firstPropertyId, lock, state );
         return this;
     }
 
     @Override
-    protected boolean loop()
+    protected boolean loadNextFromDisk()
     {
-        return !fetched;
+        return !fetched && (state == null || state.getAddedProperty( propertyKeyId ) == null);
```
Comment (ragadeeshu) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StoreSinglePropertyCursor.java]: perhaps do this in multiple lines

**Review by ragadeeshu — APPROVED** [review]
Some minor comments

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxAllPropertyCursor.java (hunk 3 above)
Comment (davidegrohmann) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/cursor/TxAllPropertyCursor.java]: This code has been removed in later commits, so there is no need to add comments.

File: community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java (hunk 2 above)
Comment (davidegrohmann) [reply community/primitive-collections/src/main/java/org/neo4j/collection/primitive/PrimitiveIntCollections.java]: Nope

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java (hunk 5 above)
Comment (davidegrohmann) [reply community/kernel/src/main/java/org/neo4j/kernel/impl/api/store/StorePropertyCursor.java]: Yes, I am gonna make the looping condition more obvious.

**Comment by davidegrohmann** [comment]
@ragadeeshu comments fixed

**Pull request closed without merging**